addons/wiki/models.py

```diff
@@ -82,6 +85,226 @@ def build_wiki_url(node, label, base, end):
+        """ The raw text of the page, suitable for using in a test search"""
+
+        return sanitize(self.html(node), tags=[], strip=True)
+
+    @property
+    def rendered_before_update(self):
+        return self.modified < WIKI_CHANGE_DATE
+
+    def get_draft(self, node):
+        """
+        Return most recently edited version of wiki, whether that is the
+        last saved version or the most recent sharejs draft.
+        """
+
+        db = wiki_utils.share_db()
+        sharejs_uuid = wiki_utils.get_sharejs_uuid(node, self.wiki_page.page_name)
+
+        doc_item = db['docs'].find_one({'_id': sharejs_uuid})
+        if doc_item:
+            sharejs_version = doc_item['_v']
+            sharejs_timestamp = doc_item['_m']['mtime']
+            sharejs_timestamp /= 1000  # Convert to appropriate units
+            sharejs_date = datetime.datetime.utcfromtimestamp(sharejs_timestamp).replace(tzinfo=pytz.utc)
+
+            if sharejs_version > 1 and sharejs_date > self.modified:
+                return doc_item['_data']
+
+        return self.content
+
+    def save(self, *args, **kwargs):
+        rv = super(WikiVersion, self).save(*args, **kwargs)
+        if self.wiki_page.node:
+            self.wiki_page.node.update_search()
+        self.wiki_page.modified = self.created
+        self.wiki_page.save()
+        self.spam_check()
+        return rv
+
+    def spam_check(self):
+        # Since wiki_pages_current will be removed from Node model, when a new WikiVersion is saved, trigger a spam check.
+        request, user_id = get_request_and_user_id()
+        request_headers = {}
+        if not isinstance(request, DummyRequest):
+            request_headers = {
+                k: v
+                for k, v in get_headers_from_request(request).items()
+                if isinstance(v, basestring)
+            }
+        user = OSFUser.load(user_id)
+        return self.wiki_page.node.check_spam(user, ['wiki_pages_latest'], request_headers)
+
+    def clone_version(self, wiki_page, user):
+        """Clone a node wiki page.
+        :param wiki_page: The wiki_page you want attached to the clone.
+        :return: The cloned wiki version
+        """
+        clone = self.clone()
+        clone.wiki_page = wiki_page
+        clone.user = user
+        clone.save()
+        return clone
+
+    @property
+    def absolute_api_v2_url(self):
+        path = '/wiki_versions/{}/'.format(self._id)
+        return api_v2_url(path)
+
+    # used by django and DRF
+    def get_absolute_url(self):
+        return self.absolute_api_v2_url
+
+
+class WikiPage(GuidMixin, BaseModel):
+    page_name = models.CharField(max_length=200, validators=[validate_page_name, ])
+    user = models.ForeignKey('osf.OSFUser', null=True, blank=True, on_delete=models.CASCADE)
+    node = models.ForeignKey('osf.AbstractNode', null=True, blank=True, on_delete=models.CASCADE, related_name='wikis')
+    deleted = NonNaiveDateTimeField(blank=True, null=True)
```

Comment:
We'll be filtering on this field frequently, so it'd be good to have an index on it.

Reply:
Note: better to do this in the migration to create the WikiPage model rather than doing this after all the WikiPage objects are created. Adding an index to a field after it's created results in a table lock unless you use `CONCURRENTLY`.